File: src/lib/queue.ts
export class Queue<T> {
  private count: number;
  private lowCount: number;
  private items: {[key: number]: T};

  constructor() {
    this.count = 0;
    this.lowCount = 0;
    this.items = {};
  }

  public enqueue(element: T): void {
    this.items[this.count] = element;
    this.count++;
  }

  public dequeue(): T | undefined {
    if (this.isEmpty()) {
      return undefined;
    }

    const item = this.items[this.lowCount];
    delete this.items[this.lowCount];
    this.lowCount++;

    return item;
  }

  public peek(): T | undefined {
    if (this.isEmpty()) return undefined;

    return this.items[this.lowCount];
  }

  public isEmpty(): boolean {
    return this.count === this.lowCount;
  }

  public size(): number {
    return this.count - this.lowCount;
  }

  public clear(): void {
    this.items = {};
    this.count = 0;
    this.lowCount = 0;
  }

  public clone(): Queue<T> {
    const queue = new Queue<T>();

    for (let i = 0; i <= this.size() - 1; i++) {
      queue.enqueue(this.items[i]);
    }

    return queue;
  }

  public to_string(): string {
    const cloned = this.clone();
    let result = '';

    for (let i = 0; (i = cloned.size()); i++) {
      result += `${cloned.dequeue()}`;
    }

    return result;
  }
}

File: lua/lib/queue.lua
Queue = {}
Queue.__index = Queue

function Queue.new()
    local self = setmetatable({}, Queue)
    self.high = 1
    self.low = 1
    self.items = {}
    return self
end

function Queue:enqueue(element)
    self.items[self.high] = element
    self.high = self.high + 1
end

function Queue:dequeue()
    if self:is_empty() then
        return nil
    end

    local item = self.items[self.low]
    self.items[self.low] = nil
    self.low = self.low + 1

    return item
end

function Queue:peek()
    return self:is_empty() and nil or self.items[self.low]
end

function Queue:size()
    return self.high - self.low
end

function Queue:is_empty()
    return self:size() <= 0
end

function Queue:clear()
    self.items = {}
    self.high = 1
    self.low = 1
end

function Queue:clone()
    local clone = Queue.new()
    clone.low = self.low
    clone.high = self.high
    clone.items = {table.unpack(self.items)}
    return clone
end

function Queue:to_string()
    local result = ""
    local cloned = Queue:clone()

    while not cloned:is_empty() do
        local item = cloned:dequeue()
        result = result .. tostring(item)
    end

    return result
end
